cmd/integration-test/pkg/tests/workload_proxy.go

```diff
@@ -0,0 +1,290 @@
+// Copyright (c) 2024 Sidero Labs, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+package tests
+
+import (
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"errors"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/cosi-project/runtime/pkg/resource/rtestutils"
+	"github.com/siderolabs/go-pointer"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/siderolabs/omni/client/pkg/client"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/omni"
+	"github.com/siderolabs/omni/cmd/integration-test/pkg/clientconfig"
+	"github.com/siderolabs/omni/internal/backend/workloadproxy"
+)
+
+// AssertWorkloadProxy tests the workload proxy feature.
+func AssertWorkloadProxy(testCtx context.Context, apiClient *client.Client, clusterName string) TestFunc {
+	return func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(testCtx, 150*time.Second)
+		t.Cleanup(cancel)
+
+		st := apiClient.Omni().State()
+		kubeClient := getKubernetesClient(ctx, t, apiClient.Management(), clusterName)
+		port := 12345
+		label := "test-nginx"
+		icon := "H4sICKNuvmYAA1NpZGVyb19tYXJrX29uZWNvbG9yX1JHQi5zdmcAfVRLaxtBDP4ry/aqkUead+oYUl/bU6CHXsqycR7gPIhNnFD63/tpvYSmpMF4RszI0vfQeLl7uuqeb7d3u9P+er9/OFksDocDHwLfP14t1Hu/QEbfPd1sDl/un0973/lO" +
```

Comment:
not a big deal, but we could probably embed the icon itself via `go:embed` and base64/gzip it in the tests